### PR TITLE
feat: CupertinoThemeExtension

### LINKED ISSUE #116876: Add extensions property to CupertinoThemeData
## Use case

The `MaterialApp` has a property `theme` that accepts `ThemeData` which has a property `extensions`, the CupertinoApp has also a property `theme` but it accepts `CupertinoThemeData` and it doesn't have the property `extensions`. I would like a solution to extend my theme the same way I can do it using Material.

## Proposal

Add the `extensions` property to `CupertinoThemeData` as well in order to use code like

`Theme.of(context).extension<MyTheme>()?.myPropertyName`

### PR DESCRIPTION
**Pull request by xsahil03x**
Introduces `CupertinoThemeExtension` for `CupertinoThemeData` similar to `ThemeExtension` available in `ThemeData`.

<details>
<summary>Code Sample</summary>

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/scheduler.dart';

/// Flutter code sample for [CupertinoThemeExtension].

const Duration kThemeChangeDuration = Duration(milliseconds: 180);

@immutable
class MyColors extends CupertinoThemeExtension<MyColors> {
  const MyColors({
    required this.brandColor,
    required this.danger,
  });

  final Color? brandColor;
  final Color? danger;

  @override
  MyColors copyWith({Color? brandColor, Color? danger}) {
    return MyColors(
      brandColor: brandColor ?? this.brandColor,
      danger: danger ?? this.danger,
    );
  }

  @override
  CupertinoThemeExtension<MyColors> resolveFrom(BuildContext context) {
    return MyColors(
      brandColor: CupertinoDynamicColor.maybeResolve(brandColor, context),
      danger: CupertinoDynamicColor.maybeResolve(danger, context),
    );
  }

  // Optional
  @override
  String toString() => 'MyColors(brandColor: $brandColor, danger: $danger)';
}

void main() {
  // Slow down time to see lerping.
  timeDilation = 5.0;
  runApp(const CupertinoThemeExtensionExampleApp());
}

class CupertinoThemeExtensionExampleApp extends StatefulWidget {
  const CupertinoThemeExtensionExampleApp({super.key});

  @override
  State<CupertinoThemeExtensionExampleApp> createState() => _CupertinoThemeExtensionExampleAppState();
}

class _CupertinoThemeExtensionExampleAppState extends State<CupertinoThemeExtensionExampleApp> {
  bool isLightTheme = true;

  void toggleTheme() {
    setState(() => isLightTheme = !isLightTheme);
  }

  @override
  Widget build(BuildContext context) {
    return CupertinoApp(
      theme: CupertinoThemeData(
        brightness: isLightTheme ? Brightness.light : Brightness.dark,
        extensions: const <CupertinoThemeExtension<dynamic>>[
          MyColors(
            brandColor: CupertinoDynamicColor.withBrightness(
              color: Color(0xFF1E88E5),
              darkColor: Color(0xFF90CAF9),
            ),
            danger: CupertinoDynamicColor.withBrightness(
              color: Color(0xFFE53935),
              darkColor: Color(0xFFEF9A9A),
            ),
          ),
        ],
      ),
      home: Home(
        isLightTheme: isLightTheme,
        toggleTheme: toggleTheme,
      ),
    );
  }
}

class Home extends StatelessWidget {
  const Home({
    super.key,
    required this.isLightTheme,
    required this.toggleTheme,
  });

  final bool isLightTheme;
  final void Function() toggleTheme;

  @override
  Widget build(BuildContext context) {
    final MyColors myColors = CupertinoTheme.of(context).extension<MyColors>()!;
    return CupertinoPageScaffold(
      navigationBar: const CupertinoNavigationBar(
        middle: Text('CupertinoThemeExtension Sample'),
      ),
      child: Center(
        child: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            AnimatedContainer(width: 100, height: 100, color: myColors.brandColor, duration: kThemeChangeDuration),
            const SizedBox(width: 10),
            AnimatedContainer(width: 100, height: 100, color: myColors.danger, duration: kThemeChangeDuration),
            const SizedBox(width: 50),
            CupertinoButton.filled(
              onPressed: toggleTheme,
              child: Icon(isLightTheme ? CupertinoIcons.moon_zzz_fill : CupertinoIcons.sun_max_fill),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

CupertinoThemeExtensionSample

Fixes: #116876

### Breaking Changes:
- `NoDefaultCupertinoThemeData.new` and `CupertinoThemData.new` is no longer a const constructor and are converted into a factory constructor for simplifying the extension input.

### Migration Guide
// TODO

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/theme.dart [hunk 1]
```diff
@@ -11,6 +11,34 @@ import 'text_theme.dart';
 
 export 'package:flutter/foundation.dart' show Brightness;
 
+/// An interface that defines custom additions to a [CupertinoThemeData] object.
+///
+/// Typically used for custom colors. To use, subclass [CupertinoThemeExtension],
+/// define a number of fields (e.g. [Color]s), and implement the [copyWith] and
+/// [lerp] methods. The latter will ensure smooth transitions of properties when
```
Comment (xsahil03x) [inline packages/flutter/lib/src/cupertino/theme.dart:18]:
```suggestion
/// [resolveFrom] methods. The latter will ensure smooth transitions of properties when
```

File: packages/flutter/lib/src/cupertino/theme.dart [hunk 2]
```diff
@@ -144,6 +172,27 @@ class _InheritedCupertinoTheme extends InheritedWidget {
...
+    // Strangely, the cast is necessary for tests to run.
+    for (final CupertinoThemeExtension<dynamic> extension in extensionsIterable) extension.type: extension as CupertinoThemeExtension<CupertinoThemeExtension<dynamic>>,
+  });
+}
+
+/// Linearly interpolate between two [extensions].
+///
+/// Includes all theme extensions in [a] and [b].
+///
+/// {@macro dart.ui.shadow.lerp}
```
Comment (xsahil03x) [inline packages/flutter/lib/src/cupertino/theme.dart:188]:
```suggestion
/// Returns a copy of the current [extensions] with all the fields resolved 
/// against the given [BuildContext].
```

**Comment by xsahil03x** [comment]
@HansMuller @MitchellGoodwin can you guys please review and suggest the required changes? Thanks.

**Comment by xsahil03x** [comment]
@TahaTesser can you help in getting this PR reviewed? Thanks.

**Comment by MitchellGoodwin** [comment]
@xsahil03x thank you for putting this together! I can help this get reviewed, but it looks like there's still some linter errors. The customer testing failing makes sense however with the breaking changes you mentioned, and Google testing is likely going to fail as well. Once this PR is ready, we'll have to figure out a migration strategy, as this will likely break a lot of tests in particular. Thank you for updating all the tests so far, it looks like it was a pain!

File: packages/flutter/lib/src/cupertino/theme.dart [hunk 3]
```diff
@@ -144,6 +172,24 @@ class _InheritedCupertinoTheme extends InheritedWidget {
...
+/// Convert the [extensionsIterable] passed to [CupertinoThemeData.new] or [copyWith]
+/// to the stored [extensions] map, where each entry's key consists of the extension's type.
+Map<Object, CupertinoThemeExtension<dynamic>> _themeExtensionIterableToMap(Iterable<CupertinoThemeExtension<dynamic>> extensionsIterable) {
+  return Map<Object, CupertinoThemeExtension<dynamic>>.unmodifiable(<Object, CupertinoThemeExtension<dynamic>>{
+    // Strangely, the cast is necessary for tests to run.
+    for (final CupertinoThemeExtension<dynamic> extension in extensionsIterable) extension.type: extension as CupertinoThemeExtension<CupertinoThemeExtension<dynamic>>,
+  });
+}
+
+/// Returns a copy of the current [extensions] with all the fields resolved
```
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/theme.dart:184]: Looks like the analyze test is complaining about a trailing space at the end of this line

Comment (xsahil03x) [reply packages/flutter/lib/src/cupertino/theme.dart]: fixed.

**Comment by xsahil03x** [comment]
@MitchellGoodwin can you review again?

**Comment by kaboc** [comment]
I found this PR while I was thinking about how to make one of my packages support ThemeExtension. Wouldn't it be better to move the existing ThemeExtension out of material to enable it to be shared between Material and Cupertino (and more varieties of designs that can be added in the future) even if the change is breaking, rather than adding CupertinoThemeExtension?

**Comment by MitchellGoodwin** [comment]
Hello 👋. Thank you very much for putting this together, I know it was a pain. This is 100% a feature we want in, but there are two issues. For one we are currently investigating making changes to themes as a whole to allow for more customizability. As such we may fulfill this need somewhere else. For another, this PR represents a large amount of breaking changes for developers that we are hesitant to introduce, especially with the ongoing investigation. As such, I am going to close this PR, but we will reopen it if we think it will fit in the future. Again, thank you very much, I know this was a pain.

**Pull request closed without merging**